rrs/tools/upgrade_history_internal.py

```diff
@@ -0,0 +1,617 @@
+    '747c7dc8702d2241475894876d06a2f1f2b29fed',
+    '369bbf393438ae4a76ab0d1817463c6f735816ea',
+    'dd5208ae22d47504443785daece4bff6331d8904',
+    'c9fdf3d046606a0becb2e6b566a481c483b9021a',
+    '13269dfbbc62faef32595343dc78250fdb2a2946',
+    '270a1e9bcf26a43f5cbdc5b901c4c6f79495311d',
+    '15e876ada73fe8e98284d14dec166007b5767f19',
+    '647db1d9eb65b225ffbb6953f796232026bfa935',
+    '75529d384bfeaf52befccb892cf41f22dc02668b',
+    'ec9fcdf14d3e2aefc5af1e53a69f056db6ea83f5',
+    '84980150ff4a2c27acd1f27123f200e03bee8c4b',
+    'c93dc7dd18a752d9523e11c6c4dce1908b5970b4',
+    '2d9a8a5539342faa1827f4902b1095a9f3448c66',
+    'b2cd021887e12d9f5b8ba48d9be3c2f2119c8e2a',
+    'da3659155cd1825a4a8d3d7c5288b4273714de15',
+    '4af90876914e5f2ccc5b7f833cd43c239c2dac55',
+    '8a771f22980f766b71f3ea0825568fc5c669e444',
+    'b0338efcdabeec79c568c74b6888d7d523e8e9dc',
+    'f3f394913b4e4a7c601ad1158faaf8b9d493e1c7',
+    'b3e246fef166030f327b5a852718ea907ada1759',
+    'a8a2c5ec891286a1e7fd5ebdd33565f9ae3965c2',
+    'ee48cb68e5d91ba108cccdabce003233290ba816',
+    '088814ef79015d9df0c8c8bc61219507cfe52ad5',
+    'c03cef42e079e4ed3d1e4f401722778157158bd6',
+    'bcdaa93dc70411da8876364ae67d0bf2456a3611',
+    'e8dfe9799e473e0ba911a0670aa23e8e8d700223',
+    'e38e56e28f2090e2b8013546f4dd76da8d59f766',
+    '85981cbbf0ce48a6d82bc39248afa9540ca858d8',
+    '147f5a665fe5073027d92e4acac631f15f08f79f',
+    'b503b1fe9a71f70726c92f46a71fc49615256fce',
+    '4972faf1bf20f07a1c1f608bc421c6fd05651594',
+    '309a02931779f32d1139cc1169a039cbe4638706',
+]
+
+
+"""
+    Store upgrade into RecipeUpgrade model.
+"""
+def _save_upgrade(recipesymbol, layerbranch, pv, srcrev, license, commit, title, info, filepath, logger, upgrade_type=None, orig_filepath=None, prev_version=None):
+    from rrs.models import Maintainer, RecipeUpgrade
+
+    maintainer_name = info.split(';')[0]
+    maintainer_email = info.split(';')[1]
+    author_date = info.split(';')[2]
+    commit_date = info.split(';')[3]
+
+    maintainer = Maintainer.create_or_update(maintainer_name, maintainer_email)
+
+    upgrade = RecipeUpgrade()
+    upgrade.recipesymbol = recipesymbol
+    upgrade.maintainer = maintainer
+    upgrade.author_date = rfc2822_time_to_utc_datetime(author_date)
+    upgrade.commit_date = rfc2822_time_to_utc_datetime(commit_date)
+    upgrade.version = pv
+    upgrade.srcrev = srcrev
+    upgrade.sha1 = commit
+    upgrade.title = title.strip()
+    upgrade.filepath = filepath
+    if upgrade_type:
+        upgrade.upgrade_type = upgrade_type
+    if orig_filepath:
+        upgrade.orig_filepath = orig_filepath
+    if prev_version:
+        upgrade.prev_version = prev_version
+    upgrade.license = license
+    upgrade.regroup()
+    upgrade.save()
+
+"""
+    Create upgrade receives new recipe_data and cmp versions.
+"""
+def _create_upgrade(recipe_data, layerbranch, ct, title, info, filepath, logger, initial=False, orig_filepath=None):
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This function calls 20 other functions. By comparison, 99% of the functions in the CodeGuru reference dataset use fewer. This indicates the function is highly coupled with other functions. A function that is highly coupled with other functions is difficult to understand and its behavior might change unexpectedly when one of its referenced functions is updated. High coupling could also increase the integration test complexity, maintenance cost and technical debt. We recommend that you simplify this function or break it into multiple functions.